talk/objectorientation/typecasting.tex

```diff
@@ -0,0 +1,66 @@
+\subsection[cast]{Type casting}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Type casting}
+  \begin{block}{5 types of casts in \cpp}
+    \begin{itemize}
+    \item \mintinline{c++}{static_cast<Target>(arg)}: Convert type at compile time if allowed by type system
+    \item \mintinline{c++}{dynamic_cast<Target>(arg)}: Check if object at address of ``\mintinline{c++}{arg}'' is compatible with the type \mintinline{c++}{Target}.
+          \alt<1>{Throw exception if it's not.}{Return \mintinline{c++}{nullptr} if it's not.}
+    \end{itemize}
+  \end{block}
+
+  \vspace{-5mm}
+
+  \begin{overprint}
+    \onslide<1>
+    \begin{exampleblock}{}
+      \begin{cppcode*}{gobble=4}
+        class A{public: virtual ~A(){}} a;
+        class B : public A { }          b;
+
+        A& c = static_cast<A&>(b); // OK. b is also an A
+        B& d = static_cast<B&>(a); // UB: a is not a B
+        B& e = static_cast<B&>(c); // OK. b is a B
+
+        B& f = dynamic_cast<B&>(c); // OK, c is a B
+        B& g = dynamic_cast<B&>(a); // Exception: not a B
+      \end{cppcode*}
+    \end{exampleblock}
+    \onslide<2>
+    \begin{exampleblock}{}
+      \begin{cppcode*}{gobble=4}
+        B* d = dynamic_cast<B*>(&a); // nullptr. a not a B.
+        if (d != nullptr) {
+          // Will not reach this
+        }
+
+        if (auto bPtr = dynamic_cast<B*>(&c)) {
+          // OK, we will get here
+        }
+      \end{cppcode*}
+    \end{exampleblock}
+  \end{overprint}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Type casting}
+  \begin{block}{5 types of casts in \cpp}
+    \begin{itemize}
+    \item \mintinline{c++}{const_cast}: Remove constness from a type. If you think you need this, first try to improve the design!
```

Comment:
```suggestion
    \item \mintinline{c++}{const_cast}: Remove constness from a type. If you think you need this, first try to improve your design!
```

Reply:
Most of the time, it wasn't "my" design. Rejected! 🙂

Reply:
fine